File: C/Consumatore.c
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#define MAX_STRING_LENGTH 256

int main(int argc, char* argv[]){

	char *file_in, read_char, *pref;
	int nread, fd;
    int check[256] = {0}; //Inizializzo un array di zero per il controllo dei caratteri da tagliare
	
	//controllo numero argomenti
	if (argc != 2 && argc != 3){ 
		perror(" numero di argomenti sbagliato"); exit(1);
	} 
	
	pref = argv[1];
    
    //Per evitare di iterare ogni volta sull'array pref per controllare se i caratteri sono presenti,
    //Inserisco 1 (vero) nelle posizioni indicate dalla codifica ASCII dei caratteri inseriti dall'utente
    
    for (int i = 0; i < strlen(pref); i++) {
        check[pref[i]] = 1;
    }
	
	//Nel caso in cui l'utente abbia inserito due argomenti, ridireziono l'input sul file indicato (se esiste)
	//Così da non dover differenziare il codice
	
	if (argc == 3) {
        file_in = argv[2];
        fd = open(file_in, O_RDONLY);
        if (fd<0){
            perror("P0: Impossibile aprire il file.");
            exit(2);
        }
        close(0);
        dup(fd);
        close(fd);
    }
	
	while (nread = read(0, &read_char, 1)) {
		if (nread >= 0) {
            if (!check[read_char]) {  //Controllo dei caratteri da filtrare
                putchar(read_char);
            }
        }
		else{
			printf("(PID %d) impossibile leggere dal file %s", getpid(), file_in);
			perror("Errore!");
			close(fd);
			exit(3);
		}
	}
}

File: Esercitazione #0/Java/Consumatore.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;

public class Consumatore {
	public static void main(String[] args) {
		BufferedReader r = null;
		char ch;
		int x;
		boolean[] check = new boolean[256]; //Inizializzato a false
		
		if (args.length != 1 && args.length != 2){
			System.out.println("Numero di argomenti errato");
			System.exit(0);
		}
		
		if (args.length == 1) {
			r = new BufferedReader(new InputStreamReader(System.in));
		} else {
			try {
				r = new BufferedReader(new InputStreamReader(new FileInputStream(args[1])));
			} catch (FileNotFoundException e) {
				System.out.println("File non trovato");
				System.exit(1);
			}
		}
		
		//Come nell'esercizio in C, per evitare di iterare ogni volta su args[0], ovvero il prefisso
		
		for (int i = 0; i < args[0].length(); i++) {
			check[(int) args[0].charAt(i)] = true;
		}
		
		try {
			while ((x = r.read()) >= 0) { 
				ch = (char) x;
				if (!check[x]) {
					System.out.print(ch);
				}
			}
			r.close();
		} catch(IOException ex){
			System.out.println("Errore di input");
			System.exit(2);
		}
}}
